flink-runtime/src/test/java/org/apache/flink/runtime/util/ZooKeeperUtilsTreeCacheTest.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.flink.shaded.curator4.org.apache.curator.framework.recipes.cache.TreeCacheListener;
+import org.apache.flink.shaded.guava18.com.google.common.io.Closer;
+
+import org.apache.curator.test.TestingServer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Tests for {@link ZooKeeperUtils#createTreeCache(CuratorFramework, String,
+ * org.apache.flink.util.function.RunnableWithException)}.
+ */
+public class ZooKeeperUtilsTreeCacheTest extends TestLogger {
+
+    private static final String PARENT_PATH = "/foo";
+    private static final String CHILD_PATH = PARENT_PATH + "/bar";
+
+    private Closer closer;
+    private CuratorFramework client;
+
+    private final AtomicReference<CompletableFuture<Void>> callbackFutureReference =
+            new AtomicReference<>();
+
+    @Before
+    public void setUp() throws Exception {
+        closer = Closer.create();
+        final TestingServer testingServer = closer.register(new TestingServer());
+
+        Configuration configuration = new Configuration();
+        configuration.set(
+                HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, testingServer.getConnectString());
+
+        client = closer.register(ZooKeeperUtils.startCuratorFramework(configuration));
+        client = closer.register(ZooKeeperUtils.startCuratorFramework(configuration));
+        final TreeCache cache =
+                closer.register(
+                        ZooKeeperUtils.createTreeCache(
+                                client,
+                                CHILD_PATH,
+                                () -> callbackFutureReference.get().complete(null)));
+        cache.start();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        closer.close();
+        callbackFutureReference.set(null);
+    }
+
+    @Test
+    public void testCallbackCalledOnNodeCreation() throws Exception {
+        client.create().forPath(PARENT_PATH);
+        callbackFutureReference.set(new CompletableFuture<>());
+        client.create().forPath(CHILD_PATH);
+        callbackFutureReference.get().get();
+    }
+
+    @Test
+    public void testCallbackCalledOnNodeModification() throws Exception {
+        testCallbackCalledOnNodeCreation();
+
+        callbackFutureReference.set(new CompletableFuture<>());
+        client.setData().forPath(CHILD_PATH, new byte[1]);
+        callbackFutureReference.get().get();
+    }
+
+    @Test
+    public void testCallbackCalledOnNodeDeletion() throws Exception {
+        testCallbackCalledOnNodeCreation();
+
+        callbackFutureReference.set(new CompletableFuture<>());
+        client.delete().forPath(CHILD_PATH);
+        callbackFutureReference.get().get();
```

Comment:
Just so I understand these tests - am I right in thinking that the tests for node creation/modification/deletion are asserted by `get()`ing the underlying `CompletableFuture` callback, and that it's completion signifies the callback was successful? If this test failed, would this method hang or is there a timeout to the `get()`?

I was wondering whether there's an equivalent `FlinkMatchers.willComplete(Duration.ofMillis(20)))` that would call the `get()` variant with a timeout, for consistency & to be clear what is being asserted?

Reply:
correct; the get asserts that the future is completed. If it fails the test will never finish. 
Recent discussions have revealed that people do not like timeouts in tests, so I'm adhering to that.

Reply:
Cool! Thanks for the clarification.

Reply:
The problem with timeouts is that if your timeout is too short for slow CI machines, then you might get false negative tests.